theano/tensor/opt.py

```diff
@@ -2591,6 +2595,139 @@ def local_adv_sub1_adv_inc_sub1(node):
+
+                # `xi` may have more dimensions than `y` since the subtensor
+                # ops do automatic broadcasting of the increment internally.
+                # Thus, we need to make the leading implicitly broadcasted
+                # dimensions explicit for shape comparison later.
+                if xi.ndim > y.ndim:
+                    y = T.shape_padleft(y, xi.ndim - y.ndim)
+                    node.fgraph.__import_r__([y], '%s: add `y`' % reason)
+
+                # Build `z_shape` and `z_broad` explicitly to include extra
+                # implicit dimensions.
+                z_shape = ((1,) * (max(xi.ndim, y.ndim) - z.ndim) +
+                           tuple(z.shape))
+                z_broad = ((True,) * (max(xi.ndim, y.ndim) - z.ndim) +
+                           z.broadcastable)
+
+                cond = [# The shapes of `y` and `xi` must either agree or `y`
+                        # may also have shape equal to 1 which may be treated
+                        # as a broadcastable dimension by the subtensor op.
+                        T.or_(T.eq(y.shape[k], xi.shape[k]),
+                              T.eq(y.shape[k], 1))
+                        # If the increment is broadcastable in dimension `k`,
+                        # we need to check whether the shapes of `xi` and `y`
+                        # are the same.
+                        if z_broad[k]
+                        # Else, we need to check whether the shapes of `xi` and
+                        # `z` are the same in dimension `k`.
+                        else T.eq(xi.shape[k], z_shape[k])
+                        # Loop over all dimensions.
+                        for k in xrange(xi.ndim)
+                        # If we can infer statically that the shape of `xi` and
+                        # `y` are the same in dimension `k` we do not need to
+                        # check it at run time.
+                        if not same_shape(xi, y, dim_x=k, dim_y=k)]
+            except AttributeError:
+                # The shape feature may not be available in some mode, but we
+                # need it for this optimization, so don't continue.
+                return False
+
+            if len(cond) > 0:
+                msg = '`x[i]` and `y` do not have the same shape.'
+                z = Assert(msg)(z, *cond)
+
+            return [node.op(x, z, *i)]
+
+
+#################
+# Bit-wise opts #
+#################
+
+@register_canonicalize
+@register_specialize
+@register_stabilize
+@gof.local_optimizer([T.Elemwise])
+def local_useless_OR(node):
+    """
+    Remove useless OR op.
+    """
+    if (isinstance(node.op, T.Elemwise) and
+        isinstance(node.op.scalar_op, scalar.basic.OR)):
+        assert len(node.inputs) == 2
+        for i in node.inputs:
+            try:
+                if get_scalar_constant_value(i) != 0:
+                    return [T.as_tensor_variable(1)]
```

Comment:
This optimization is startd in https://github.com/Theano/Theano/issues/2160. It only miss tests and someone work on it. I think it would be better to remove it from your PR to not duplicate work and allow faster merging of this PR.